e2e/consultemplate/consultemplate.go

```diff
@@ -30,123 +33,231 @@ func init() {
-		messages := []string{}
-		for _, event := range events {
-			messages = append(messages, event.DisplayMessage)
-		}
+	// verify we've rendered the templates
+	for allocID := range downstreams {
+		f.NoError(waitForTemplateRender(allocID, "task/local/kv.yml",
+			func(out string) bool {
+				return strings.TrimSpace(out) == "---\nkey: foo\njob: templating"
+			}, nil), "expected consul key to be rendered")
 
-		return messages
-	}, 5*time.Second, time.Second).Should(ContainElement(ContainSubstring("kv.block")))
+		f.NoError(waitForTemplateRender(allocID, "task/local/services.conf",
+			func(out string) bool {
+				confLines := strings.Split(strings.TrimSpace(out), "\n")
+				servers := 0
+				for _, line := range confLines {
+					if strings.HasPrefix(line, "server upstream-service ") {
+						servers++
+					}
+				}
+				return servers == 2
+			}, nil), "expected 2 upstream servers")
+	}
 
-	// Insert consultemplatetest
-	_, err = consulClient.KV().Put(&capi.KVPair{Key: "consultemplatetest", Value: []byte("bar")}, nil)
-	require.Nil(err)
+	// Update our key in Consul
+	_, err = tc.Consul().KV().Put(&capi.KVPair{Key: key, Value: []byte("bar")}, nil)
+	f.NoError(err)
 
-	// Placement should start running
-	waitForClientAllocStatus(structs.AllocClientStatusRunning)
+	// Wait for restart
+	for allocID, group := range downstreams {
+		var checkErr error
+		testutil.WaitForResultRetries(retries, func() (bool, error) {
+			time.Sleep(interval)
+			out, err := e2e.Command("nomad", "alloc", "status", allocID)
+			f.NoError(err, "could not get allocation status")
 
-	// Ensure restart count 0 -- we should be going from blocked to running.
-	waitForRestartCount(0)
+			section, err := e2e.GetSection(out, "Task Events:")
+			f.NoError(err, out)
 
-	// Update consultemplatetest
-	_, err = consulClient.KV().Put(&capi.KVPair{Key: "consultemplatetest", Value: []byte("baz")}, nil)
-	require.Nil(err)
+			restarts, err := e2e.GetField(section, "Total Restarts")
+			f.NoError(err)
+			return restarts == "1",
+				fmt.Errorf("expected 1 restart for %q but found %s", group, restarts)
+		}, func(e error) {
+			checkErr = e
+		})
+		f.NoError(checkErr)
 
-	// Wrap in retry to wait until restart
-	waitForRestartCount(1)
-}
+		// verify we've re-rendered the template
+		f.NoError(waitForTemplateRender(allocID, "task/local/kv.yml",
+			func(out string) bool {
+				return strings.TrimSpace(out) == "---\nkey: bar\njob: templating"
+			}, nil), "expected updated consul key")
+	}
 
-func (tc *ConsulTemplateTest) AfterEach(f *framework.F) {
-	nomadClient := tc.Nomad()
-	consulClient := tc.Consul()
+	// increase the count for upstreams
+	count := 3
+	job.TaskGroups[2].Count = &count
+	_, _, err = tc.Nomad().Jobs().Register(job, nil)
+	f.NoError(err, "could not register job")
 
-	jobs := nomadClient.Jobs()
-	// Stop all jobs in test
-	for _, id := range tc.jobIds {
-		jobs.Deregister(id, true, nil)
+	// wait for re-rendering
```

Comment:
do we want a test that `noop` is honored?

Reply:
Good call.

Reply:
Fixed in 4fa768d